compiler/rustc_typeck/src/collect/object_lifetime_defaults.rs

```diff
@@ -0,0 +1,543 @@
+
+impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
+    fn with<F>(&mut self, wrap_scope: Scope<'_>, f: F)
+    where
+        F: for<'b> FnOnce(&mut LifetimeContext<'b, 'tcx>),
+    {
+        let LifetimeContext { tcx, defs, .. } = self;
+        let mut this = LifetimeContext { tcx: *tcx, defs, scope: &wrap_scope };
+        f(&mut this);
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    fn visit_segment_args(
+        &mut self,
+        res: Res,
+        depth: usize,
+        generic_args: &'tcx hir::GenericArgs<'tcx>,
+    ) {
+        if generic_args.parenthesized {
+            for input in generic_args.inputs() {
+                self.visit_ty(input);
+            }
+            let output = generic_args.bindings[0].ty();
+            self.visit_ty(output);
+            return;
+        }
+
+        let tcx = self.tcx;
+
+        // Figure out if this is a type/trait segment,
+        // which requires object lifetime defaults.
+        let type_def_id = match res {
+            Res::Def(DefKind::AssocTy, def_id) if depth == 1 => Some(tcx.parent(def_id)),
+            Res::Def(DefKind::Variant, def_id) if depth == 0 => Some(tcx.parent(def_id)),
+            Res::Def(
+                DefKind::Struct
+                | DefKind::Union
+                | DefKind::Enum
+                | DefKind::TyAlias
+                | DefKind::Trait,
+                def_id,
+            ) if depth == 0 => Some(def_id),
+            _ => None,
+        };
+        debug!(?type_def_id);
+
+        if let Some(type_def_id) = type_def_id {
+            // Compute a vector of defaults, one for each type parameter,
+            // per the rules given in RFCs 599 and 1156. Example:
+            //
+            // ```rust
+            // struct Foo<'a, T: 'a, U> { }
+            // ```
+            //
+            // If you have `Foo<'x, dyn Bar, dyn Baz>`, we want to default
+            // `dyn Bar` to `dyn Bar + 'x` (because of the `T: 'a` bound)
+            // and `dyn Baz` to `dyn Baz + 'static` (because there is no
+            // such bound).
+            //
+            // Therefore, we would compute `object_lifetime_defaults` to a
+            // vector like `['x, 'static]`. Note that the vector only
+            // includes type parameters.
+            let generics = tcx.generics_of(type_def_id);
+
+            let in_body = {
+                let mut scope = self.scope;
+                loop {
+                    match *scope {
+                        Scope::Root => break false,
+                        Scope::Body => break true,
+                        Scope::Binder { s, .. }
+                        | Scope::Static { s, .. }
+                        | Scope::ObjectLifetimeDefault { s, .. } => {
+                            scope = s;
+                        }
+                    }
+                }
+            };
+            let object_lifetime_default = |i: usize| {
+                let param = generics.params.get(i)?;
```

Comment:
This is wrong.  The index of argument in HIR may not correspond to the index in `ty::Generics`.